common/logger/src/sample.rs

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implement periodic sampling for logs
+
+use std::{
+    sync::atomic::{AtomicU64, Ordering},
+    time::{Duration, SystemTime},
+};
+
+#[derive(Debug)]
+pub enum SampleRate {
+    Duration(Duration),
+    Frequency(u64),
+    Always,
+}
+
+pub struct Sampling {
+    rate: SampleRate,
+    state: AtomicU64,
+}
+
+impl Sampling {
+    pub const fn new(rate: SampleRate) -> Self {
+        Self {
+            rate,
+            state: AtomicU64::new(0),
+        }
+    }
+
+    pub fn sample(&self) -> bool {
+        match &self.rate {
+            SampleRate::Duration(rate) => Self::sample_duration(rate, &self.state),
+            SampleRate::Frequency(rate) => Self::sample_frequency(*rate, &self.state),
+            SampleRate::Always => true,
+        }
+    }
+
+    fn sample_frequency(rate: u64, count: &AtomicU64) -> bool {
+        let previous_count = count
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |count| {
+                let new_count = if count == 0 {
+                    rate.saturating_sub(1)
+                } else {
+                    count.saturating_sub(1)
+                };
+                Some(new_count)
+            })
+            .expect("Closure should always returns 'Some'. This is a Bug.");
+
+        previous_count == 0
+    }
+
+    fn sample_duration(rate: &Duration, last_sample: &AtomicU64) -> bool {
+        let rate = rate.as_secs();
+        // Seconds since Unix Epoch
+        let now = SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .expect("SystemTime before UNIX EPOCH!")
+            .as_secs();
+
+        last_sample
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |last_sample| {
+                if now.saturating_sub(last_sample) >= rate {
+                    Some(now)
+                } else {
+                    None
+                }
+            })
+            .is_ok()
+    }
+}
+
+#[macro_export]
+macro_rules! sample {
+    ($sample_rate:expr, $($args:expr)+ ,) => {
+        $crate::sample!($sample_rate, $($args)+);
+    };
+
+    ($sample_rate:expr, $($args:tt)+) => {{
+        static SAMPLING: Sampling = Sampling::new($sample_rate);
+        if SAMPLING.sample() {
+            $($args)+
+        }
+    }};
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn frequency() {
+        // Frequency
+        let sampling = Sampling::new(SampleRate::Frequency(10));
+        let mut v = Vec::new();
+        for i in 0..=25 {
```

Comment:
nit: this is the only test that uses `0..=##` vs the others use `0..##`.  Though, I just learned about this now!